File: testdata/fixtures/csharp/CWE-89.cs
using System.Data.SqlClient;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace Fixture {
    public class OrderController : Controller {
        private readonly AppDb _db;

        public IActionResult ByName() {
            string name = Request.Query["name"];
            // ruleid: csharp-sql-built-from-request
            var cmd = new SqlCommand("SELECT * FROM Orders WHERE Name = '" + name + "'", _conn);
            cmd.ExecuteNonQuery();
            return Ok();
        }

        public IActionResult ByCity(string city) {
            // ruleid: csharp-sql-built-from-request
            var cmd = new SqlCommand($"SELECT * FROM Orders WHERE City = '{city}'", _conn);
            cmd.ExecuteReader();
            return Ok();
        }

        public IActionResult ByRegion() {
            string region = Request.Form["region"];
            var text = "SELECT * FROM Orders WHERE Region = '" + region + "'";
            // ruleid: csharp-sql-built-from-request
            _db.Orders.FromSqlRaw(text);
            return Ok();
        }

        public IActionResult Purge(string tenant) {
            // ruleid: csharp-sql-built-from-request
            _db.Database.ExecuteSqlRaw("DELETE FROM Orders WHERE Tenant = '" + tenant + "'");
            return Ok();
        }

        public IActionResult Parameterised(string city) {
            // ok: csharp-sql-built-from-request
            var cmd = new SqlCommand("SELECT * FROM Orders WHERE City = @city", _conn);
            cmd.Parameters.AddWithValue("@city", city);
            return Ok();
        }

        public IActionResult Constant() {
            // ok: csharp-sql-built-from-request
            var cmd = new SqlCommand("SELECT * FROM Orders", _conn);
            return Ok();
        }

        public IActionResult Internal() {
            string table = Config.TableName;
            // ok: csharp-sql-built-from-request
            var cmd = new SqlCommand("SELECT * FROM " + table, _conn);
            return Ok();
        }

        private SqlConnection _conn;
    }
}

File: testdata/fixtures/kotlin/CWE-89.kt
// CWE-89: an SQL statement is assembled from a string an app did not choose.
//
// On Android the caller is another app on the phone. An intent's extras are
// whatever the sender put in them, and `SELECT * FROM users WHERE name = '$n'`
// with `n` set to `' OR '1'='1` reads the whole table. The exported activity
// that takes the intent is the front door; nothing in between validates it.
package fixture

import android.app.Activity
import android.content.Intent
import android.database.sqlite.SQLiteDatabase

class Orders : Activity() {

    private lateinit var db: SQLiteDatabase

    fun byName(intent: Intent) {
        val name = intent.getStringExtra("name")
        // ruleid: kotlin-sql-built-from-untrusted-input
        db.rawQuery("SELECT * FROM orders WHERE name = '" + name + "'", null)
    }

    fun byCity(intent: Intent) {
        val city = intent.getStringExtra("city")
        // ruleid: kotlin-sql-built-from-untrusted-input
        db.execSQL("DELETE FROM orders WHERE city = '$city'")
    }

    fun byRegion(intent: Intent) {
        val region = intent.data.toString()
        val text = "SELECT * FROM orders WHERE region = '" + region + "'"
        // ruleid: kotlin-sql-built-from-untrusted-input
        db.rawQuery(text, null)
    }

    fun byTenant(tenant: String) {
        // ok: kotlin-sql-built-from-untrusted-input
        db.rawQuery("SELECT * FROM orders WHERE tenant = ?", arrayOf(tenant))
    }

    fun everything() {
        // ok: kotlin-sql-built-from-untrusted-input
        db.rawQuery("SELECT * FROM orders", null)
    }

    fun byConfigured() {
        val table = Config.tableName
        // ok: kotlin-sql-built-from-untrusted-input
        db.rawQuery("SELECT * FROM " + table, null)
    }
}
